File: example/src/example/Predef.py
from typing import Optional

import AST
import TypeSystem as TS
from AST import SymbolRef
from Result import Result, Warn, Success, Failure
from TypeSystem import VarArg, Type


def prepare(dict_: dict[str, Type]) -> dict[str, SymbolRef]:
    return {name: SymbolRef(name, None, type_) for name, type_ in dict_.items()}


unary_numerical_type = TS.Function(TS.Int(), TS.Int()) | TS.Function(TS.Float(), TS.Float())
unary_vector_type = TS.FunctionTypeFactory(
    args=TS.Vector(),
    result_hint=TS.Vector(),
    result_type_factory=lambda expr: Success(expr.type)
)
unary_matrix_type = TS.FunctionTypeFactory(
    args=TS.Matrix(),
    result_hint=TS.Matrix(),
    result_type_factory=lambda expr: Success(expr.type)
)

binary_numerical_type = TS.Function((TS.Int(), TS.Int()), TS.Int()) \
                        | TS.Function((TS.numerical(), TS.numerical()), TS.Float())

binary_numerical_condition_type = TS.Function((TS.numerical(), TS.numerical()), TS.Bool())


def binary_metrix_type_factory(first: AST.Expr[TS.Matrix], second: AST.Expr[TS.Matrix]) -> Result[TS.Type]:
    a, b = first.type, second.type
    assert isinstance(a, TS.Matrix) and isinstance(b, TS.Matrix)

    errors = []
    warns = []

    rows: Optional[int] = None
    cols: Optional[int] = None

    if a.rows is None or b.rows is None:
        warns.append("Matrix rows could not be inferred")
    elif a.rows != b.rows:
        errors.append(f"Matrix rows mismatch: {a.rows} != {b.rows}")
    else:
        rows = a.rows

    if a.cols is None or b.cols is None:
        warns.append("Matrix columns could not be inferred")
    elif a.cols != b.cols:
        errors.append(f"Matrix columns mismatch: {a.cols} != {b.cols}")
    else:
        cols = a.cols

    if errors:
        return Failure(TS.Matrix(rows, cols), "\n".join(errors))
    elif warns:
        return Warn(TS.Matrix(rows, cols), "\n".join(warns))
    else:
        return Success(TS.Matrix(a.rows, a.cols))


def binary_vector_type_factory(first: AST.Expr[TS.Vector], second: AST.Expr[TS.Vector]) -> Result[TS.Type]:
    a, b = first.type, second.type
    assert isinstance(a, TS.Vector) and isinstance(b, TS.Vector)
    if a.arity is None or b.arity is None:
        return Warn(TS.Matrix(), "Vector arity could not be inferred")
    elif a.arity != b.arity:
        return Failure(TS.Matrix(), f"Vector lengths mismatch: {a.arity} != {b.arity}")
    else:
        return Success(TS.Matrix(a.arity, a.arity))


binary_matrix_type = TS.FunctionTypeFactory(
    args=(TS.Matrix(), TS.Matrix()),
    result_hint=TS.Matrix(),
    result_type_factory=binary_metrix_type_factory
)

binary_vector_type = TS.FunctionTypeFactory(
    args=((TS.Vector()), (TS.Vector())),
    result_hint=(TS.Vector()),
    result_type_factory=binary_vector_type_factory
)

scalar_type = TS.FunctionTypeFactory(
    args=(TS.Matrix(), TS.numerical()),
    result_hint=TS.Matrix(),
    result_type_factory=lambda expr, args: Success(expr.type)
) | TS.FunctionTypeFactory(
    args=(TS.Vector(), TS.numerical()),
    result_hint=TS.Vector(),
    result_type_factory=lambda expr, args: Success(expr.type)
)


def matrix_create_function_type(size: AST.Expr[TS.Int]) -> Result[TS.Type]:
    match size:
        case AST.Literal(n):
            return Success(TS.Matrix(n, n))
        case _:
            return Warn(TS.Matrix(), "Matrix size could not be inferred")


matrix_type = TS.FunctionTypeFactory(
    args=TS.Int(),
    result_hint=TS.Matrix(),
    result_type_factory=matrix_create_function_type
)

unary = prepare({
    "UMINUS": unary_numerical_type | unary_vector_type | unary_matrix_type,
    "'": TS.FunctionTypeFactory(
        args=(TS.Matrix()),
        result_hint=(TS.Matrix()),
        result_type_factory=lambda expr: Success(TS.Matrix(expr.type.cols, expr.type.rows))
    ),
    "eye": matrix_type,
    "zeros": matrix_type,
    "ones": matrix_type,
})

binary = prepare({
    "+": binary_numerical_type,
    "-": binary_numerical_type,
    "*": binary_numerical_type | scalar_type | binary_matrix_type | TS.Function((TS.String(), TS.Int()), TS.String()),
    "/": binary_numerical_type | scalar_type,
    "==": binary_numerical_condition_type,
    "!=": binary_numerical_condition_type,
    "<=": binary_numerical_condition_type,
    ">=": binary_numerical_condition_type,
    ">": binary_numerical_condition_type,
    "<": binary_numerical_condition_type,
    ".+": binary_matrix_type | binary_vector_type,
    ".-": binary_matrix_type | binary_vector_type,
    ".*": binary_matrix_type | binary_vector_type,
    "./": binary_matrix_type | binary_vector_type,
})


def init_vector_factory(*args: AST.Expr[TS.Vector]) -> Result[TS.Type]:
    # assert all(isinstance(arg, TS.Vector) for arg in arg_types)
    arities = set(arg.type.arity for arg in args)  # type: ignore
    if len(arities) == 1:
        return Success(TS.Matrix(len(args), arities.pop()))
    else:
        if None not in arities:
            return Warn(TS.Matrix(len(args)), f"Vector arities {arities} are not the same")
        return Warn(TS.Matrix(), "Cannot infer matrix size")


var_args = prepare({
    "INIT": TS.FunctionTypeFactory(
        args=VarArg(TS.numerical()),
        result_hint=TS.Vector(),
        result_type_factory=lambda *args: Success(TS.Vector(len(args)))
    ) | TS.FunctionTypeFactory(
        args=VarArg(TS.Vector()),
        result_hint=TS.Matrix(),
        result_type_factory=init_vector_factory
    ),

    "PRINT": TS.Function(VarArg(TS.Any()), TS.unit()),
})

symbols: dict[str, SymbolRef] = {**unary, **binary, **var_args}


# todo: maybe split into two functions?
def get_symbol(name: str) -> SymbolRef:
    res = symbols[name]
    if isinstance(res, SymbolRef):
        return res.copy()
    else:
        return res

File: example/src/example/Predef.scala
package example

import java.security.DrbgParameters.Reseed
import scala.collection.mutable

val unary_numerical_type = Type.Function(Tuple(Type.Int), Type.Int) | Type.Function(Tuple(Type.Float), Type.Float)
val unary_vector_type = Type.FunctionTypeFactory(
  arg = Type.Vector(),
  resultHint = Type.Vector(),
  resultTypeFactory = expr => Result.Success(expr.tpe),
)
val unary_matrix_type = Type.FunctionTypeFactory(
  arg = Type.Matrix(),
  resultHint = Type.Matrix(),
  resultTypeFactory = expr => Result.Success(expr.tpe),
)

val binary_numerical_type = Type.Function((Type.Int, Type.Int), Type.Int) |
  Type.Function((Type.Numerical, Type.Numerical), Type.Float)

val binary_numerical_condition_type = Type.Function((Type.Numerical, Type.Numerical), Type.Bool)

val binary_matrix_type = Type.FunctionTypeFactory(
  args = (Type.Matrix(), Type.Matrix()),
  resultHint = Type.Matrix(),
  resultTypeFactory =
    case (AST.Expr(a: Type.Matrix), AST.Expr(b: Type.Matrix)) =>
      val errors = mutable.Set.empty[String]
      val warns = mutable.Set.empty[String]

      var rows: Int | Null = null
      var cols: Int | Null = null

      if (a.rows == null) || (b.rows == null) then warns += "Matrix rows could not be inferred"
      else if a.rows != b.rows then errors += s"Matrix rows mismatch: ${a.rows} != ${b.rows}"
      else rows = a.rows

      if (a.cols == null) || (b.cols == null) then warns += "Matrix columns could not be inferred"
      else if a.cols != b.cols then errors += s"Matrix columns mismatch: ${a.cols} != ${b.cols}"
      else cols = a.cols

      if errors.nonEmpty then Result.error(Type.Matrix(rows, cols), errors.toSeq*)
      else if warns.nonEmpty then Result.warn(Type.Matrix(rows, cols), warns.toSeq*)
      else Result.Success(Type.Matrix(a.rows, a.cols))
    case _ => Result.error(Type.Matrix(), "Invalid arguments"),
)

val binary_vector_type = Type.FunctionTypeFactory(
  args = (Type.Vector(), Type.Vector()),
  resultHint = Type.Vector(),
  resultTypeFactory =
    case (AST.Expr(a: Type.Vector), AST.Expr(b: Type.Vector)) =>
      if (a.arity == null) || (b.arity == null) then Result.warn(Type.Matrix(), "Vector arity could not be inferred")
      else if a.arity != b.arity then Result.error(Type.Matrix(), s"Vector lengths mismatch: ${a.arity} != ${b.arity}")
      else Result.Success(Type.Matrix(a.arity, a.arity))
    case _ => Result.error(Type.Matrix(), "Invalid arguments"),
)

val scalar_type = Type.FunctionTypeFactory(
  args = (Type.Matrix(), Type.Numerical),
  resultHint = Type.Matrix(),
  resultTypeFactory =
    case (expr, args) => Result.Success(expr.tpe),
) | Type.FunctionTypeFactory(
  args = (Type.Vector(), Type.Numerical),
  resultHint = Type.Vector(),
  resultTypeFactory =
    case (expr, args) => Result.Success(expr.tpe),
)

val matrix_type = Type.FunctionTypeFactory(
  arg = Type.Int,
  resultHint = Type.Matrix(),
  resultTypeFactory =
    case AST.Literal(_, n: Int, _) => Result.Success(Type.Matrix(n, n))
    case _ => Result.warn(Type.Matrix(), "Matrix size could not be inferred"),
)

val symbols: Map[String, AST.SymbolRef] = Map(
  // unary
  "UMINUS" -> (unary_numerical_type | unary_vector_type | unary_matrix_type),
  "TRANSPOSE" -> Type.FunctionTypeFactory(
    arg = Type.Matrix(),
    resultHint = Type.Matrix(),
    resultTypeFactory = expr =>
      val tpe = expr.tpe.asInstanceOf[Type.Matrix]
      Result.Success(Type.Matrix(tpe.cols, tpe.rows)),
  ),
  "EYE" -> matrix_type,
  "ZEROS" -> matrix_type,
  "ONES" -> matrix_type,

  // binary
  "+" -> binary_numerical_type,
  "-" -> binary_numerical_type,
  "*" -> (binary_numerical_type | scalar_type | binary_matrix_type | Type.Function(Tuple(Type.String), Type.Int)),
  "/" -> (binary_numerical_type | scalar_type),
  "==" -> binary_numerical_condition_type,
  "!=" -> binary_numerical_condition_type,
  "<=" -> binary_numerical_condition_type,
  ">=" -> binary_numerical_condition_type,
  ">" -> binary_numerical_condition_type,
  "<" -> binary_numerical_condition_type,
  "DOTADD" -> (binary_matrix_type | binary_vector_type),
  "DOTSUB" -> (binary_matrix_type | binary_vector_type),
  "DOTMUL" -> (binary_matrix_type | binary_vector_type),
  "DOTDIV" -> (binary_matrix_type | binary_vector_type),

  // varargs
  "INIT" ->
    (Type.FunctionTypeFactory.varargs(
      arg = Type.VarArg(Type.Numerical),
      resultHint = Type.Vector(),
      resultTypeFactory = args => Result.Success(Type.Vector(args.size)),
    ) | Type.FunctionTypeFactory.varargs(
      arg = Type.VarArg(Type.Vector()),
      resultHint = Type.Matrix(),
      resultTypeFactory = args =>
        args.map(_.tpe.asInstanceOf[Type.Vector].arity).distinct match
          case Seq(arity) =>
            Result.Success(Type.Matrix(args.size, arity))
          case arities =>
            if !arities.contains(null) then
              Result.warn(Type.Matrix(args.size), f"Vector arities {arities} are not the same")
            else Result.warn(Type.Matrix(), "Cannot infer matrix size"),
    )),
  "PRINT" -> Type.Function(Tuple(Type.VarArg(Type.Any)), Type.Unit),
).map:
  case (name, tpe) => name -> AST.SymbolRef(tpe, name, -1)
